src/clp_ffi_py/ir/PyDecoderBuffer.hpp

```diff
@@ -0,0 +1,178 @@
+        m_input_ir_stream = nullptr;
+    }
+
+    /**
+     * Releases the memory allocated for underlying metadata field and the
+     * reference hold for the Python object(s).
+     */
+    auto clean() -> void {
+        Py_XDECREF(m_input_ir_stream);
+        PyMem_Free(m_read_buffer_mem_owner);
+    }
+
+    /**
+     * Cleans the consumed bytes by shifting the unconsumed bytes to the
+     * beginning of the buffer, and fills the read buffer by reading from the
+     * input IR stream. If more than half of the bytes are unconsumed in the
+     * read buffer, the buffer will be doubled before reading.
+     * @param num_bytes_read Number of bytes read from the input IR stream to
+     * populate the read buffer.
+     * @return true on success.
+     * @return false on failure with the relevant Python exception and error
+     * set.
+     */
+    [[nodiscard]] auto populate_read_buffer(Py_ssize_t& num_bytes_read) -> bool;
+
+    /**
+     * Commits the bytes consumed in the read buffer by incrementing the
+     * underlying cursor position.
+     * @param num_bytes_consumed Total number of bytes consumed.
+     * @return true on success.
+     * @return false on failure with the relevant Python exception and error
+     * set.
+     */
+    auto commit_read_buffer_consumption(Py_ssize_t num_bytes_consumed) -> bool;
+
+    [[nodiscard]] auto get_num_decoded_message() const -> size_t { return m_num_decoded_message; }
+
+    auto increment_num_decoded_message() -> void { ++m_num_decoded_message; }
+
+    /**
+     * @return Number of unconsumed bytes stored in the current read buffer.
+     */
+    [[nodiscard]] auto get_num_unconsumed_bytes() const -> Py_ssize_t {
+        return m_buffer_size - m_num_current_bytes_consumed;
+    }
+
+    /**
+     * @return A span containing unconsumed bytes.
+     */
+    [[nodiscard]] auto get_unconsumed_bytes() const -> gsl::span<int8_t> {
+        return m_read_buffer.subspan(m_num_current_bytes_consumed, get_num_unconsumed_bytes());
+    }
+
+    /**
+     * Handles the Python buffer protocol's `getbuffer` operation.
+     * This function should fail unless the buffer protocol is enabled.
+     * @param view Python Buffer view.
+     * @param flags Python Buffer flags.
+     * @return 0 on success.
+     * @return -1 on failure with the relevant Python exception and error set.
+     */
+    [[nodiscard]] auto py_getbuffer(Py_buffer* view, int flags) -> int;
+
+    /**
+     * @return Whether the buffer protocol is enabled.
+     */
+    [[nodiscard]] auto is_py_buffer_protocol_enabled() const -> bool {
+        return m_py_buffer_protocol_enabled;
+    }
+
+    /**
+     * Tests the functionality of the DecoderBuffer by stepping through the
+     * input stream and stream it as random sized byte sequences. It will
+     * attempt to read a random number of bytes from the read buffer and grow
+     * the buffer when necessary until the the entire input stream is consumed.
```

Comment:
```suggestion
     * Tests the functionality of the DecoderBuffer by sequentially reading
     * through the input stream with randomly sized reads. It will grow the read
     * buffer when necessary until the the entire input stream is consumed.
```